src/ipset/ipset_hash_net.c

```diff
@@ -0,0 +1,121 @@
+#include "ipset/ipset_hash.h"
+#include "ipset/pfxlen.h"
+
+typedef struct hash_net_elem {
+    union inet_addr ip;
+    uint8_t cidr;
+    int proto;
+    char comment[IPSET_MAXCOMLEN];
+} elem_t;
+
+static bool 
+hash_net_data_equal4(const void *elem1, const void *elem2)
+{
+    elem_t *e1 = (elem_t *)elem1;
+    elem_t *e2 = (elem_t *)elem2;
+
+    return inet_addr_equal(AF_INET, &e1->ip, &e2->ip) && 
+           e1->cidr == e2->cidr;
+}
+
+static void
+hash_net_do_list4(struct ipset_member *member, void *elem, bool comment)
+{
+    elem_t *e = (elem_t *)elem;
+
+    member->addr.in.s_addr = e->ip.in.s_addr;
+    member->cidr = e->cidr;    
+    member->proto = e->proto;
+    if (comment)
+        rte_strlcpy(member->comment, e->comment, IPSET_MAXCOMLEN);
+}
+
+static uint32_t
+hash_net_hashkey4(void *data, int len, uint32_t mask)
+{
+    elem_t *e = (elem_t *)data;
+
+    return (e->ip.in.s_addr * 31 + e->cidr * 31 + e->proto * 31) & mask;
+}
+
+static int
+hash_net_adt4(int op, struct ipset *set, struct ipset_param *param)
+{
+    elem_t e;
+    int ret = 0;
+    uint32_t ip, ip_to;
+    ipset_adtfn adtfn = set->type->adtfn[op];
+
+    if (set->family != param->option.family)
+        return EDPVS_INVAL;
+
+    memset(&e, 0, sizeof(e));
+
+    e.cidr = param->cidr;
+    e.proto = param->proto;
+
+    if (op == IPSET_OP_TEST) {
+        e.ip.in.s_addr = param->range.min_addr.in.s_addr;
+
+        return adtfn(set, &e);
+    }
+
+    ip = ntohl(param->range.min_addr.in.s_addr);
+
+    if (e.cidr || (!e.cidr && param->range.max_addr.in.s_addr == 0)) {
+        ip_set_mask_from_to(ip, ip_to, e.cidr);
```

Comment:
If the second condition stands `!e.cidr && param->range.max_addr.in.s_addr == 0`, then ip=0 after `ip_set_mask_from_to(ip, ip_to, e.cidr)`. Is it correct?

Reply:
range.min_addr won't be larger than range.max_addr